src/threadpoolwork-inl.h

```diff
@@ -24,13 +24,16 @@
 
 #if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
 
-#include "util-inl.h"
 #include "node_internals.h"
+#include "tracing/trace_event.h"
+#include "util-inl.h"
 
 namespace node {
 
 void ThreadPoolWork::ScheduleWork() {
   env_->IncreaseWaitingRequestCounter();
+  TRACE_EVENT_NESTABLE_ASYNC_BEGIN0(
+      TRACING_CATEGORY_NODE1(threadpool), type_name(), this);
   int status = uv_queue_work(
       env_->event_loop(),
       &work_req_,
```

Comment:
What do you think about tracing sync events? Trace analyzer can compute thread CPU usage from synchronous events. For thread poll work, I believe it can be more helpful to track the synchronous work instead of the nestable asynchronous IO.

Reply:
I think users care more about the total time. But it also makes sense to track the time of synchronous operations. We need to use flow event if we need to trace all of this ?

Reply:
Sounds good to me!

Reply:
The code as follow does not work. I load the trace event data to tracer viewer and the panel is empty. So I'm still using `NESTABLE_ASYNC_EVENT` to trace the event.
```
void ThreadPoolWork::ScheduleWork() {
  env_->IncreaseWaitingRequestCounter();
  TRACE_EVENT_FLOW_BEGIN0(
      TRACING_CATEGORY_NODE1(threadpoolwork), type_name(), this);
  int status = uv_queue_work(
      env_->event_loop(),
      &work_req_,
      [](uv_work_t* req) {
        ThreadPoolWork* self = ContainerOf(&ThreadPoolWork::work_req_, req);
        TRACE_EVENT_FLOW_STEP0(
          TRACING_CATEGORY_NODE1(threadpoolwork), self->type_name(), self, "start");
        self->DoThreadPoolWork();
        TRACE_EVENT_FLOW_STEP0(
          TRACING_CATEGORY_NODE1(threadpoolwork), self->type_name(), self, "end");
      },
      [](uv_work_t* req, int status) {
        ThreadPoolWork* self = ContainerOf(&ThreadPoolWork::work_req_, req);
        self->env_->DecreaseWaitingRequestCounter();
        TRACE_EVENT_FLOW_END1(TRACING_CATEGORY_NODE1(threadpoolwork),
                                        self->type_name(),
                                        self,
                                        "result",
                                        status);
        self->AfterThreadPoolWork(status);
      });
  CHECK_EQ(status, 0);
}
```

Reply:
That doesn't seem to work either. I can't find any use cases about flow event.

Reply:
@theanarkh shouldn't we do the `TRACE_EVENT_BEGIN0()` call inside the `work_cb` lambda instead of the function that's calling `uv_queue_work()`? Both the `TRACE_EVENT_BEGIN0()` call and the `TRACE_EVENT_END0()` calls need to be on the same thread for it to work properly.

Reply:
The code as follows does not work.
```
void ThreadPoolWork::ScheduleWork() {
  env_->IncreaseWaitingRequestCounter();
  TRACE_EVENT_FLOW_BEGIN0(
      TRACING_CATEGORY_NODE1(threadpoolwork), type_name(), this);
  int status = uv_queue_work(
      env_->event_loop(),
      &work_req_,
      [](uv_work_t* req) {
        ThreadPoolWork* self = ContainerOf(&ThreadPoolWork::work_req_, req);
        TRACE_EVENT_BEGIN0(...);
        self->DoThreadPoolWork();
        TRACE_EVENT_END0(...);
      },
      [](uv_work_t* req, int status) {
        ThreadPoolWork* self = ContainerOf(&ThreadPoolWork::work_req_, req);
        self->env_->DecreaseWaitingRequestCounter();
        TRACE_EVENT_FLOW_END1(TRACING_CATEGORY_NODE1(threadpoolwork),
                                        self->type_name(),
                                        self,
                                        "result",
                                        status);
        self->AfterThreadPoolWork(status);
      });
  CHECK_EQ(status, 0);
}
```
Do you have any suggestions ? How about this ?
```
void ThreadPoolWork::ScheduleWork() {
  env_->IncreaseWaitingRequestCounter();
  TRACE_EVENT_NESTABLE_ASYNC_BEGIN0(
      TRACING_CATEGORY_NODE1(threadpoolwork), type_, this);
  int status = uv_queue_work(
      env_->event_loop(),
      &work_req_,
      [](uv_work_t* req) {
        ThreadPoolWork* self = ContainerOf(&ThreadPoolWork::work_req_, req);
        TRACE_EVENT_NESTABLE_ASYNC_INSTANT0(
            TRACING_CATEGORY_NODE1(threadpoolwork), self->type_, self);
        self->DoThreadPoolWork();
        TRACE_EVENT_NESTABLE_ASYNC_INSTANT0(
            TRACING_CATEGORY_NODE1(threadpoolwork), self->type_, self);
      },
      [](uv_work_t* req, int status) {
        ThreadPoolWork* self = ContainerOf(&ThreadPoolWork::work_req_, req);
        self->env_->DecreaseWaitingRequestCounter();
        TRACE_EVENT_NESTABLE_ASYNC_END1(TRACING_CATEGORY_NODE1(threadpoolwork),
                                        self->type_,
                                        self,
                                        "result",
                                        status);
        self->AfterThreadPoolWork(status);
      });
  CHECK_EQ(status, 0);
}
```

Reply:
@theanarkh I've suggested the change I was referring to (the test passes locally with this addition).

Reply:
Sorry, but I don't understand what you mean. Can you provide the code of `ScheduleWork`, thanks !

Reply:
@theanarkh sure :)

<details>

```cc
void ThreadPoolWork::ScheduleWork() {
  env_->IncreaseWaitingRequestCounter();
  int status = uv_queue_work(
      env_->event_loop(),
      &work_req_,
      [](uv_work_t* req) {
        ThreadPoolWork* self = ContainerOf(&ThreadPoolWork::work_req_, req);
        TRACE_EVENT_BEGIN0(TRACING_CATEGORY_NODE1(threadpoolwork), self->type_);
        self->DoThreadPoolWork();
      },
      [](uv_work_t* req, int status) {
        ThreadPoolWork* self = ContainerOf(&ThreadPoolWork::work_req_, req);
        self->env_->DecreaseWaitingRequestCounter();
        TRACE_EVENT_END0(TRACING_CATEGORY_NODE1(threadpoolwork), self->type_);
        self->AfterThreadPoolWork(status);
      });
  CHECK_EQ(status, 0);
}
```

</details>

Reply:
I think users care more about the total time instead of the time of synchronous operations ? If we also want to track the time of synchronous operations. I think the code is as follows.
```
void ThreadPoolWork::ScheduleWork() {
  env_->IncreaseWaitingRequestCounter();
 // time1
  int status = uv_queue_work(
      env_->event_loop(),
      &work_req_,
      [](uv_work_t* req) {
        ThreadPoolWork* self = ContainerOf(&ThreadPoolWork::work_req_, req);
        // time2
        self->DoThreadPoolWork();
        // time3
      },
      [](uv_work_t* req, int status) {
        ThreadPoolWork* self = ContainerOf(&ThreadPoolWork::work_req_, req);
        self->env_->DecreaseWaitingRequestCounter();
        self->AfterThreadPoolWork(status);
        // time4
      });
  CHECK_EQ(status, 0);
}
```